src/GafferUI/Handle.cpp

```diff
@@ -556,18 +558,62 @@ Handle::AngularDrag::AngularDrag( bool processModifiers )
 }
 
 Handle::AngularDrag::AngularDrag( const Gadget *gadget, const Imath::V3f &origin, const Imath::V3f &axis0, const Imath::V3f axis1, const DragDropEvent &dragBeginEvent, bool processModifiers )
-	:	m_rotation( 0.0f ),
+	:	m_gadget( gadget ),
+		m_rotation( 0.0f ),
 		m_axis0( axis0 ),
 		m_axis1( axis1 ),
 		m_processModifiers( processModifiers ),
 		m_preciseMotionEnabled( false )
 {
 	// We need to negate this, or rotation is opposite to the mouse movement direction
 	V3f planeAxis0 = -axis0.cross( axis1 );
-	// Disable modifier processing as we'll do our own precision mode in angle space
-	m_drag = PlanarDrag( gadget, origin, planeAxis0, axis1, dragBeginEvent, false );
 
-	m_dragBeginRotation = closestRotation( m_drag.startPosition(), m_rotation );
+	WorldGadgetDragData dragData;
+	worldGadgetDragData( gadget, origin, planeAxis0, axis1, dragBeginEvent, dragData );
+
+	if( abs( dragData.worldPlane.normal.dot( dragData.worldLine.dir ) ) < g_planarToLinearDragThreshold )
+	{
+		// Do a LinearDrag in raster space instead of world space
+
+		float axis0Dot = dragData.worldAxis0.dot( dragData.worldLine.dir );
+		float axis1Dot = dragData.worldAxis1.dot( dragData.worldLine.dir );
+
+		bool useAxis0 = abs( axis0Dot ) < abs( axis1Dot );
+
+		// Looking down the normal of the rotation plane, positive values rotate
+		// counterclockwise. Find the gadget-space vector pointing in the direction
+		// of positive rotation to use for the start direction of the linear drag.
+		V3f parallelAxis = useAxis0 ? axis1 : planeAxis0;
+		float parallelDot = useAxis0 ? axis1Dot : axis0Dot;
+
+		// If the parallel axis is pointing in the same direction as the view,
+		// reverse the drag direction.
+		float sign = -( ( 0 < parallelDot ) - ( parallelDot < 0 ) );
+
+		V3f dragDirection = planeAxis0.cross( m_axis1 ).cross( parallelAxis );
```

Comment:
I haven't fully wrapped my head around this, but if useAxis0 is true, this ends up setting drag direction to `-axis0.cross( axis1 ).cross( axis1 ).cross( axis1 )`. I feel like there's probably a way of expressing this without crossing with the same vector 3 times?

Reply:
Very nice, it turns out `planeAxis0.cross( m_axis1 )` is the same as `axis0`, so we can drop one or two crosses with that. Changed in https://github.com/GafferHQ/gaffer/pull/5281/commits/7696ac7a7149d6896d8a6c011448e00dae815329